### PR TITLE
Allow granular progress updates & associated changes

### PR DESCRIPTION
**Pull request by gselzer**
This PR refactors the scijava-progress module API, for the goals of simplicity and better progress reporting within dependencies, and should be squeaked in before the 1.0.0 release.

Important changes:
* Progress stages have been done away with, with `Task`s instead having a single tracker of `elements`, discrete packets of progress updates. The number of elements a Task will have is dictated during `Progress.defineTotal`, meaning we no longer need the `Progress.setStageMax` method.
* The progress of Op dependencies was previously limited to a boolean value, which is not helpful for "wrapper Ops", where progress updates are most important in dependency subtasks. Dependency progress updates are now passed along to *every* parent using a new method `Task.parent()`, meaning that if an Op dependency has a dependency itself that issues an update, `ProgressListeners` tuned to its parent and grandparent will be notified as well. To prevent an overwhelming amount of dependencies from flooding the SciJava Ops Legacy progress listener only posts updates for tasks with no parent.
  * To enable progress updates from Op dependencies, you must add the `"progress.TRACK"` hint (i.e. `org.scijava.ops.engine.BaseOpHints.Progress.TRACK`), **and** use `Progress.defineTotal(x, y)`, where `x` is a long defining the number of elements in the *parent* Op and `y` is the total number of reporting dependency executions. E.g.
```java
public class ParentOp implements Function<X, Y> {
  @OpDependency(name="foo.bar", hints={"progress.TRACK"})
  Function<X, Y> foo;
  @OpDependency(name="baz.bar", hints={"progress.TRACK"})
  Function<X, Y> baz;

  @Override
  public Y apply(X in) {
    Progress.defineTotal(0, 3);
    out1 = foo.apply(in);
    out2 = foo.apply(in);
    out3 = baz.apply(in);
    return out3;
  }
}
```
* Two of the most intensive Ops, `coloc.saca.heatmapZScore` and `deconvolve.richardsonLucy`, now report progress.

As we are approaching the initial release, we should consider the API - I know @ctrueden wanted to do an API review, so I'm pinging his review as motivation to review the scijava-progress public API. I'm also asking for @elevans' opinion as an Ops developer/user - if you'd like to try your hand at adding progress to fully test the API, I'd appreciate it, but if you're busy no worries.

Points of Discussion:

* Can we make it easier for progress listeners to receive updates only about top-level Ops (i.e. not Op dependencies)?
* It would be slick to add progress updates to `LoopBuilder` constructs such that pixel-wise Ops always report progress when lifted to image level, however this anecdotally incurs very poor performance due to the constant multithreaded updates to the `AtomicLong` tracking progress. @tpietzsch is there anything that we could do, here or upstream in ImgLib2, to enable progress reporting? Potentially one update per chunk would be awesome - would it be possible to add API to report chunk size, for example?

TODO:

* [x] Update documentation
* [x] Add progress to any more Ops?

### CODE REVIEW

**Review by gselzer — COMMENTED** [review]
Here are my suggestions! Thanks @gselzer for doing this work!

File: scijava-progress/src/main/java/org/scijava/progress/Progress.java [hunk 1]
```diff
@@ -332,6 +330,16 @@ public void update(long numElements) {
 			// NB: No-op
 		}
 
+		@Override
+		public double progress() {
+			return NOP_PROGRESS;
+		}
+
+		@Override
+		public void define(final long elements, final long subTasks) {
```
[the remark this answers is not in the document]
Comment (gselzer) [reply scijava-progress/src/main/java/org/scijava/progress/Progress.java]: Fixed!

File: scijava-progress/src/main/java/org/scijava/progress/Task.java [hunk 2]
```diff
@@ -142,12 +140,12 @@ public synchronized Task createSubtask( //
 	 * {@link Task#progress()} are called, as otherwise total progress cannot be
 	 * defined.
 	 * <p>
-	 * Under the hood, this method calls {@link #defineTotalProgress(long, long)}
+	 * Under the hood, this method calls {@link #define(long, long)}
 	 *
-	 * @param totalStages the number of computation stages within the task
+	 * @param elements the number of discrete "elements" in the computation
```
[the remark this answers is not in the document]
Comment (gselzer) [reply scijava-progress/src/main/java/org/scijava/progress/Task.java]: Fixed

File: scijava-progress/src/main/java/org/scijava/progress/Task.java [hunk 3]
```diff
@@ -157,16 +155,26 @@ public void defineTotalProgress(final long totalStages) {
...
-		this.numStages = totalStages;
-		this.numSubTasks = totalSubTasks;
+	public void define(final long elements, final long subTasks) {
+		if (tasksDefined) {
+			throw new IllegalStateException(
+				"Progress has already been defined for this task");
+		}
+		// Total tasks = all subtasks, plus one iff there are also progress elements
+		// in this stage
+		this.tasks = subTasks + (elements == 0 ? 0 : 1);
```
[the remark this answers is not in the document]
Comment (gselzer) [reply scijava-progress/src/main/java/org/scijava/progress/Task.java]: Folded!

File: scijava-progress/src/main/java/org/scijava/progress/Task.java [hunk 4]
```diff
@@ -109,15 +102,20 @@ public Task(final Object progressible, //
...
+			}
+			else {
+				msg += " has subtasks that did not complete!";
+			}
+			throw new IllegalStateException(msg);
+		}
 		this.completed = true;
+		if (hasElements) {
+			this.current.set(this.max.get());
+		}
```
[the remark this answers is not in the document]
Comment (gselzer) [reply scijava-progress/src/main/java/org/scijava/progress/Task.java]: Fixed!

File: scijava-progress/src/main/java/org/scijava/progress/Task.java [hunk 5]
```diff
@@ -283,4 +266,13 @@ public void update(final long numElements) {
 	public Object progressible() {
 		return progressible;
 	}
+
+	/**
+	 * Returns the parent {@link Task} of this {@code Task}.
+	 *
+	 * @return the parent {@link Task}, or {@code null} if this Task has no parent
+	 */
+	public Task parent() {
```
[the remark this answers is not in the document]
Comment (gselzer) [reply scijava-progress/src/main/java/org/scijava/progress/Task.java]: Changed!

File: scijava-ops-legacy/src/main/java/org/scijava/legacy/service/DefaultOpEnvironmentService.java [hunk 6]
```diff
@@ -86,6 +87,7 @@ public SciJavaProgressListener(TaskService tasks) {
 
 		@Override
 		public void acknowledgeUpdate(Task task) {
+			if (task.parent() != null) return;
```
[the remark this answers is not in the document]
Comment (gselzer) [reply scijava-ops-legacy/src/main/java/org/scijava/legacy/service/DefaultOpEnvironmentService.java]: Added a new method `task.isSubTask` so we don't have to expose the parent for this use case.

File: scijava-ops-tutorial/src/main/java/org/scijava/ops/tutorial/ReportingProgress.java [hunk 7]
```diff
@@ -79,18 +79,20 @@ public class ReportingProgress implements OpCollection {
...
-		Progress.defineTotalProgress(1, 0);
+		// If you have N discrete packets of computation, you should call
+		// Progress.defineTotalProgress(N)
+		// If you have N discrete packets of computation and want to call Op
+		// dependencies M times, you should call
+		// Progress.defineTotalProgress(N, M)
+
+		// Here, we want to update the progress every time we find a new prime.
+		// We call no Op dependencies, thus the call looks like:
+		Progress.defineTotal(numPrimes, 0);
```
[the remark this answers is not in the document]
Comment (gselzer) [reply scijava-ops-tutorial/src/main/java/org/scijava/ops/tutorial/ReportingProgress.java]: Much better now!

File: docs/ops/doc/WritingYourOwnOpPackage.md [hunk 8]
```diff
@@ -276,6 +276,66 @@ class DoubleSizeOp implements Function<double[], Double> {
...
+    public Double apply(final double[] inArray) {
+        // There's no significant work here, but we do have 2 subtasks.
+        Progress.defineTotal(0, 2);
+        final Double sum = sumOp.apply(inArray);
+        final Double size = sizeOp.apply(inArray);
+        return sum / size;
+    }
+}
+```
+
```
Comment (elevans) [inline docs/ops/doc/WritingYourOwnOpPackage.md:338]: Can you discuss what, if any, performance hit a developer might expect with `Progress`? Maybe a little section on ideal use versus what a **bad** use of `Progress` looks like. I don't think we need to go crazy, but just a flavor.

**Review by elevans — CHANGES_REQUESTED** [review]
@gselzer Nice PR! Ty for this work adding progress updates to SciJava Ops. This is much appreciated for those long Ops. There a few documentation changes I'd like to see before we merge this into `main`. I also have some open questions that I'd like to get your thoughts on:

1. The progress bar/feedback only really works with Fiji correct? Is there something we can do for users that are using SciJava Ops from Python or some other non-Fiji entry-point?
2. The `create.kernelDiffraction` Op that creates the PSF for deconvolution takes a little bit of time (not much but its detectable). That might be another good Op to add progress to.
3. How does this progress mechanism work with external libraries?

**Comment by gselzer** [comment]
@elevans thanks for the feedback!

>     1. The progress bar/feedback only really works with Fiji correct? Is there something we can do for users that are using SciJava Ops from Python or some other non-Fiji entry-point?

There is nothing Fiji-specific about scijava-progress. e.g. Python users could just write a new `ProgressListener` implementation and add it as a global listener, and it would receive progress updates in the same way that the Fiji progress bar does.

...this makes me think that we should probably do away with the `ProgressListener` interface, and operate on `Consumer<Task>`s - it would be a little less code for us.

>     2. The `create.kernelDiffraction` Op that creates the PSF for deconvolution takes a little bit of time (not much but its detectable). That might be another good Op to add progress to.

Sure! Do you want to try it?

>     3. How does this progress mechanism work with external libraries?

I'm not really sure what you mean by this. If you're talking about progress updates of libraries that don't depend upon scijava-progress, all Ops record binary progress updates. This means that when you invoke the Op, all `ProgressListener`s listening to that Op will get a ping that the invocation is 0% complete. Then all computation happens, and once that finishes all `ProgressListener`s will get a ping that the invocation is 100% complete. In practice, you'll just see that it's running, but you won't get granular updates until you depend on scijava-progress and do the updates described in the documentation. I don't know of a way to do any better.

If you're talking about how external libraries can register their own `ProgressListener`s, they'll naturally have to depend upon scijava-progress and create their own `ProgressListener`s.

Does that help?

File: docs/ops/doc/WritingYourOwnOpPackage.md [hunk 9]
```diff
@@ -276,6 +276,66 @@ class DoubleSizeOp implements Function<double[], Double> {
 
 ```
 
+### Defining Op Progress
+
+By adding the `scijava-progress` module, your Op can define and update its progress, providing user value for long-running Ops. To add progress to your Op, you must add the following steps to your Op:
```
[the remark this answers is not in the document]
Comment (gselzer) [reply docs/ops/doc/WritingYourOwnOpPackage.md:281]: They'd receive no feedback, unless they write their own `ProgressListener` and register it within the `Progress` class.

Naively speaking, I'd assume any scijava-ops-python component could harness [tqdm](https://tqdm.github.io/) to record progress updates, for example.

File: docs/ops/doc/WritingYourOwnOpPackage.md [hunk 10]
```diff
@@ -276,6 +276,66 @@ class DoubleSizeOp implements Function<double[], Double> {
...
+            Progress.update();
+        }
+        return i;
+    }
+}
+```
+
+If your Op defines Op dependencies whose progress you'd like to include in your total progress, you can make the following changes.
+* For each Op dependency that you want to track, pass the Hint `"progress.TRACK"` within the `@OpDependency` annotation.
+* Replace `Progress.defineTotal(long elements)` with `Progress.defineTotal(long elements, long subTasks)`, where `subTasks` is the **total** number of times you will invoke Op dependencies annotated with `"progress.TRACK"`.
```
[the remark this answers is not in the document]
Comment (gselzer) [reply docs/ops/doc/WritingYourOwnOpPackage.md:307]: Dependencies do *not* need to define their progress for updates in the dependent Op. If they do not define progress, they will only update the parent progress upon their completion, though.

For example, suppose you have 2 elements worth of updates in your Op, and in between them your Op calls a dependency, which under the hood has 10 elements:
* If that Op reports its progress i.e. calls `Progress.upate()`, the dependent Op's progress will go `0.0` -> `0.33` -> `0.36` -> `0.39` -> ... -> `0.63` -> `0.66` -> `1.0`
* If that Op **does not** report its progress the dependent Op's progress will go `0.0` -> `0.33` -> `0.66` -> `1.0`

**Comment by elevans** [comment]
Awesome thanks @gselzer for the explanations. I'm happy that you've addressed all my comments. Ready to merge!

**Pull request merged**